pvanalytics/quality/irradiance.py

```diff
@@ -1 +1,333 @@
 """Quality control functions for irradiance data."""
+
+import numpy as np
+from pvlib.tools import cosd
+
+
+QCRAD_LIMITS = {'ghi_ub': {'mult': 1.5, 'exp': 1.2, 'min': 100},
+                'dhi_ub': {'mult': 0.95, 'exp': 1.2, 'min': 50},
+                'dni_ub': {'mult': 1.0, 'exp': 0.0, 'min': 0},
+                'ghi_lb': -4, 'dhi_lb': -4, 'dni_lb': -4}
+
+QCRAD_CONSISTENCY = {
+    'ghi_ratio': {
+        'low_zenith': {
+            'zenith_bounds': [0.0, 75],
+            'ghi_bounds': [50, np.Inf],
+            'ratio_bounds': [0.92, 1.08]},
+        'high_zenith': {
+            'zenith_bounds': [75, 93],
+            'ghi_bounds': [50, np.Inf],
+            'ratio_bounds': [0.85, 1.15]}},
+    'dhi_ratio': {
+        'low_zenith': {
+            'zenith_bounds': [0.0, 75],
+            'ghi_bounds': [50, np.Inf],
+            'ratio_bounds': [0.0, 1.05]},
+        'high_zenith': {
+            'zenith_bounds': [75, 93],
+            'ghi_bounds': [50, np.Inf],
+            'ratio_bounds': [0.0, 1.10]}}}
+
+
+def _check_limits(val, lb=None, ub=None, lb_ge=False, ub_le=False):
+    """Return True where lb < (or <=) val < (or <=) ub."""
+    if lb_ge:
+        lb_op = np.greater_equal
+    else:
+        lb_op = np.greater
+    if ub_le:
+        ub_op = np.less_equal
+    else:
+        ub_op = np.less
+
+    if (lb is not None) & (ub is not None):
+        return lb_op(val, lb) & ub_op(val, ub)
+    elif lb is not None:
+        return lb_op(val, lb)
+    elif ub is not None:
+        return ub_op(val, ub)
+    else:
+        raise ValueError('must provide either upper or lower bound')
+
+
+def _qcrad_ub(dni_extra, sza, lim):
+    cosd_sza = cosd(sza)
+    cosd_sza[cosd_sza < 0] = 0
+    return lim['mult'] * dni_extra * cosd_sza**lim['exp'] + lim['min']
+
+
+def check_ghi_limits_qcrad(ghi, solar_zenith, dni_extra, limits=None):
+    r"""Test for physical limits on GHI using the QCRad criteria.
+
+    Test passes if a value > lower bound and value < upper bound. Lower bounds
+    are constant for all tests. Upper bounds are calculated as
+
+    .. math::
+        ub = min + mult * dni\_extra * cos( solar\_zenith)^{exp}
+
+    Parameters
+    ----------
+    ghi : Series
+        Global horizontal irradiance in :math:`W/m^2`
+    solar_zenith : Series
+        Solar zenith angle in degrees
+    dni_extra : Series
+        Extraterrestrial normal irradiance in :math:`W/m^2`
+    limits : dict, default QCRAD_LIMITS
+        for keys 'ghi_ub', 'dhi_ub', 'dni_ub', value is a dict with
+        keys {'mult', 'exp', 'min'}. For keys 'ghi_lb', 'dhi_lb', 'dni_lb',
+        value is a float.
+
+    Returns
+    -------
+    ghi_limit_flag : Series
+        True if value passes physically-possible test
+
+    """
+    if not limits:
+        limits = QCRAD_LIMITS
+    ghi_ub = _qcrad_ub(dni_extra, solar_zenith, limits['ghi_ub'])
+
+    ghi_limit_flag = _check_limits(ghi, limits['ghi_lb'], ghi_ub)
+    ghi_limit_flag.name = 'ghi_limit_flag'
```

Comment:
I'm not sure how valuable the `Series.name` attribute is.

Reply:
I don't think it is particularly valuable in this context. I will remove the names. (I'll also need to modify the tests to pass `check_names=False` to `assert_series_equal`, since I believe the column name is used as the series name when you get a series out of a dataframe).

Reply:
This change make sense here in a general purpose library so that you can minimize the surface of your API. But, for context, these names are used in SFA to simplify the process of combining many series into a dataframe. If that kind of combine many masks together code is in scope of `pvanalytics` then you may want to rethink the change.

Reply:
That's a very good point @wholmgren I was not aware of how the Series names were being used.

Reply:
I think at this point in development we should aim to minimize API surface. I'm further inclined to remove the names because it is difficult to know what a "good" name should be. In particular, if we end up with several different algorithms for generating a particular mask should the name include an identifier for the algorithm that was used? (I'm thinking of inverter clipping especially - I have seen three different methods for detecting this feature in different places and we may want to include several options in pvanalytics.)

If we return named series we would need to require that different implementations of the same filter return the same name or all implementations return different names (ie. by appending `_qcrad` to the name). I think both alternatives are error-prone and introduce unnecessary technical debt.

I think I see a few paths forward (but there are probably more):
* Leave it to the users who need names to add them.
* Defer naming series until a higher-level API exists for pvanalytics and add names at that level. (I'm thinking of the API for pvlib which uses a keyword argument to access different implementations.)
* Add a keyword arg to the filtering/masking functions that allows users to specify a name (but defaults to `None`). For the most part this could be accomplished with minimal changes via a decorator, but there are some filtering functions that return tuples which could complicate the decorator approach.

I prefer the first option, but if it will make the library more difficult to use then we shouldn't do it. I don't think it adds much overhead for SFA since I believe you will need to define wrapper functions anyway to apply the `mask_flags` decorator.

@wholmgren, I really appreciate your insight as a potential future user of this library.